lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -253,23 +255,179 @@ struct BinaryOpLowering : public OpConversionPattern<BinOp> {
+};
+
+HWArithToHWTypeConverter::HWArithToHWTypeConverter() {
+  // Pass any type through the signedness remover.
+  addConversion([this](Type type) { return removeSignedness(type); });
+}
+
+// Adds the ArgResOpConversion for 'TOp' to the set of conversion patterns, as
+// well as a legality check on the conversion status of the op's operands and
+// results.
+template <typename TOp>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext());
+  target.addDynamicallyLegalOp<TOp>([&](TOp op) {
+    return !typeConverter.hasSignednessSemantics(op->getOperandTypes()) &&
+           !typeConverter.hasSignednessSemantics(op->getResultTypes());
+  });
+}
+
+template <typename TOp, typename TOp2, typename... OpTs>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  addOperandConversion<TOp>(target, patterns, typeConverter);
+  addOperandConversion<TOp2, OpTs...>(target, patterns, typeConverter);
+}
+
 } // namespace
 
 //===----------------------------------------------------------------------===//
 // Pass driver
 //===----------------------------------------------------------------------===//
 
 namespace {
+
 class HWArithToHWPass : public HWArithToHWBase<HWArithToHWPass> {
 public:
   void runOnOperation() override {
     ModuleOp module = getOperation();
 
     ConversionTarget target(getContext());
     RewritePatternSet patterns(&getContext());
+    HWArithToHWTypeConverter typeConverter;
+    mlir::populateFunctionOpInterfaceTypeConversionPattern<hw::HWModuleOp>(
+        patterns, typeConverter);
 
     target.addIllegalDialect<HWArithDialect>();
     target.addLegalDialect<comb::CombDialect, hw::HWDialect>();
+    target.addDynamicallyLegalOp<hw::HWModuleOp, hw::HWModuleExternOp>(
+        [&](auto moduleLikeOp) {
+          // Legal if all results and args have no signedness integers.
+          bool legalResults = !typeConverter.hasSignednessSemantics(
+              cast<FunctionOpInterface>(moduleLikeOp).getResultTypes());
+          bool legalArgs = !typeConverter.hasSignednessSemantics(
+              cast<FunctionOpInterface>(moduleLikeOp).getArgumentTypes());
+          return legalResults && legalArgs;
+        });
+
+    // Generic conversion and legalization patterns for operations that we
+    // expect to be using in conjunction with the signedness values of hwarith.
+    addOperandConversion<
+        hw::OutputOp, comb::MuxOp, seq::CompRegOp, hw::ArrayCreateOp,
+        hw::ArrayGetOp, hw::ArrayConcatOp, hw::ArraySliceOp, hw::StructCreateOp,
+        hw::StructExplodeOp, hw::StructExtractOp, hw::StructInjectOp>(
```

Comment:
Just skimming the ops of HW and Comb, this should hopefully meet most use cases. Any particular reason why unions (`UnionCreateOp`, `UnionExtractOp`) are not included here? Or do you think we should start with a small set and extend once required?

Reply:
No particular reason, mainly due to them being used so rarely that I forgot about them!